### ISSUE TITLE
array operations too slow

### ISSUE THREAD

**Issue by oguz-ismail** [raygard/wak#2]
See:
```
$ seq 1000000 >x
$ for awk in gawk mawk bwkawk 'busybox awk' goawk wak; do echo $awk; time $awk '{$0 in m; m[$0]; delete m[$0]}' x; done
gawk

real    0m0.672s
user    0m0.671s
sys     0m0.000s
mawk

real    0m0.237s
user    0m0.227s
sys     0m0.010s
bwkawk

real    0m0.666s
user    0m0.666s
sys     0m0.000s
busybox awk

real    0m0.911s
user    0m0.901s
sys     0m0.010s
goawk

real    0m0.142s
user    0m0.140s
sys     0m0.000s
wak

real    0m3.730s
user    0m3.720s
sys     0m0.010s
```
I presume the culprit is the underlying hashmap implementation. Is there a particular reason why wak doesn't use bucket hashmaps (array of lists of elements with the same hash)? Or is it something else?

**Comment by raygard** [comment]
How did you find my implementation? I didn't think anyone would stumble across this.
Try replacing the program with e.g. '{1==1}' and show me those times. (My version of bwkawk gets a syntax error with just '{1}'.)
I think the slowness is mainly in the i/o. I'll try to make it faster, after I fix a lot of other stuff first. 
BTW your gawk seems quite faster than mine.

**Comment by oguz-ismail** [comment]
Thanks for the reply.
> How did you find my implementation? I didn't think anyone would stumble across this.`

https://lobste.rs/s/rpqo8w/awk_implementation
>Try replacing the program with e.g. '{1==1}' and show me those times.

```
$ for awk in gawk mawk bwkawk 'busybox awk' goawk wak; do echo $awk; time $awk '{1==1}' x; done
gawk

real    0m0.090s
user    0m0.089s
sys     0m0.000s
mawk

real    0m0.034s
user    0m0.034s
sys     0m0.000s
bwkawk

real    0m0.150s
user    0m0.149s
sys     0m0.000s
busybox awk

real    0m0.405s
user    0m0.405s
sys     0m0.000s
goawk

real    0m0.072s
user    0m0.056s
sys     0m0.019s
wak

real    0m3.307s
user    0m3.307s
sys     0m0.000s
```
>(My version of bwkawk gets a syntax error with just '{1}'.)

Mine too.
> I think the slowness is mainly in the i/o. I'll try to make it faster, after I fix a lot of other stuff first.

Dunno. I was trying wak with my solutions to this year's AoC puzzles, [day 23](https://adventofcode.com/2023/day/23) took too long and I thought the hashmap implementation was to blame. Both the script and input are here https://gist.github.com/oguz-ismail/e985900b4d8530f12e54c1349f5cd654 gawk, mawk, bwkawk and goawk finish under 2 minutes, wak and busybox awk take longer than 10 minutes.
> BTW your gawk seems quite faster than mine.

It's gawk 5.3.0 on Ubuntu 22.04.

**Comment by iriswebb** [comment]
TL;DR: overhead in interpx(), no idea how to reduce.
```
OPERATION  TIME (s)  TIMES USED  AVERAGE TIME (µs)
tkassign   2.481174  2000000     1.24059
drop       3.600398  1000000     3.6004
tkeq       2.185214  1000000     2.18521
```

More complex explanation:

Out of interest, I `strace`-d the programs to see the read sizes

nawk and gawk read 4096 bytes at a time, mawk reads _262143_ bytes at a time. While busybox awk only reads 462 bytes at a time. Strangely enough wak also reads 4096 bytes at a time. I doubt it's the hash table lookups since wak is slow by the same proportion when not doing them.

Testing on a debug build and dumping the bytecode shows:
```
14 zcode entries
zcode   0:   1 tkeof
zcode   1:   5 tknumber  1  1.000000
zcode   3:   5 tknumber  2  1.000000
zcode   5:  32 ==
zcode   6:  96 drop
zcode   7: 112 quit
zcode   8:  -1 NOTOKEN
zcode   9:   5 tknumber  3  123.000000
zcode  11:  60 tkexit
zcode  12: 112 quit
zcode  13: 112 quit
```
The "123" is the seed for the srandom() stuff, but debug-printf-ing through, we aren't seeding the random number generator every line.

The only operations that are running are "drop" "tkassign" and "tkeq". I timed these and averaged them with awk (Table at top, tested with a debug build made with "cc *.c -lm").

**Comment by oguz-ismail** [comment]
@oliverkwebb It's twice as faster with these changes
``` diff
diff --git a/src/lib.c b/src/lib.c
index a116bb1..392cda2 100644
--- a/src/lib.c
+++ b/src/lib.c
@@ -95,7 +95,6 @@ EXTERN char *xstrdup(char *s)
 #endif  // FOR_TOYBOX
 EXTERN double str_to_num(char *s)
 {
-  setlocale(LC_NUMERIC, "");
   return atof(s);
 }
 
diff --git a/src/main.c b/src/main.c
index 4a2581d..2ace15f 100644
--- a/src/main.c
+++ b/src/main.c
@@ -125,6 +125,7 @@ int main(int argc, char **argv)
     uselocale(h ? h : newlocale(LC_CTYPE_MASK, "en_US.UTF-8", 0));
   }
 #endif
+	setlocale(LC_NUMERIC, "");
 
   retval = awk(sepstring, progstring, prog_args, assign_args, optind,
        argc, argv, opt_run_prog);
diff --git a/src/run.c b/src/run.c
index b8ddef8..0fb7412 100644
--- a/src/run.c
+++ b/src/run.c
@@ -16,7 +16,6 @@ static void check_numeric_string(struct zvalue *v)
     // Significant speed gain with this test:
     // num string must begin space, +, -, ., or digit.
     if (strchr("+-.1234567890 ", *s)) {
-      setlocale(LC_NUMERIC, "");
       double num = strtod(s, &end);
       if (s == end || end[strspn(end, " ")]) return;
       v->num = num;
```
wak wastes too much time compiling the same regex over and over as well. Perhaps it's wiser to compile `RS` only after it's changed.

**Comment by raygard** [comment]
@oliverkwebb, thanks for the timing info etc.

@oguz-ismail, Thanks. I have the setlocale() calls on my todo list, need to
be sure how to safely move them. They are where they are from long ago when
I first worked on conversions based on what the posix spec says.

I thought I was already cacheing the FS regex, I’ll look into that. I know
I’m not yet cacheing the RS regex, and the whole regex-as-RS thing needs
work, especially in that I still use getdelim() in the code that reads
“multiline records” (still not sure that’s correct, either). I think
there’s probably some way to roll up the data-reading code to be faster and
more compact.

Now that the code is “out there” I’m getting advice faster than I can keep
up, but I’ll try to stay on top of it. (I was going to give it all a break
for a while but stuff happened.) I was (am?) more interested in getting
actual bugs out before worrying much more about speed.

Still need to get my testing regime cleaned up and posted in a repo.

Ray Gardner
Centennial, CO
raygard.net

**Comment by iriswebb** [comment]
> Thanks. I have the setlocale() calls on my todo list, need to be sure how to safely move them. They are where they are from long ago when I first worked on conversions based on what the posix spec says.

Would initializing them in `main()` be enough? I'm not too familiar with the locale subsystem but from a quick look at the man page, doing `setlocale(LC_*, "")` sets the local based on environment variables. These environment variables changing during the running of awk is _possible_, since we have the ENVIRON variable, but I don't see why it's a greater priority then being able to process lines quickly.

From what I know, LC_NUMERIC exists because of commas and decimal points in numbers, you can't put commas in numbers in awk code: "1.200" could be read different on a locale and be processed as 1200 (or maybe just 1 because strtowhatever() doesn't process separators that aren't the decimal seperator), but "1,200" gives a syntax error. Meaning you can't declare fractional numbers anymore on certain locales.

gawk doesn't seem to care, I can't find a way to configure my locale so it'll use something other than "." for a decimal separator, dunno what the solution to this is.

**Comment by raygard** [comment]
I have removed the two `setlocale()` calls from the conversion locations and put it into `awk()` so it gets called in both toybox and non-toy versions.

**Comment by raygard** [comment]
@oguz-ismail You were right, the problem was with the hash table being too slow. Try it now. I had wrongly assumed deletions would be rare but your code does a lot of them. When a table entry is deleted, I mark the slot deleted. Now, when I am inserting, I use the first slot I find that is empty *or* is marked deleted. I think this works correctly, and much faster for your case. There are still patterns of insert/delete that may slow down, but this should fix cases like yours.

You asked " Is there a particular reason why wak doesn't use bucket hashmaps (array of lists of elements with the same hash)? " I think open addressing is generally considered faster and I think is more space-efficient, and buckets with chaining have problems with deletions as well.

**Comment by oguz-ismail** [comment]
Thanks, with these changes wak is now as fast as bwkawk in all my tests except the ones that stress the regex engine. Other awks overcome the limitations of OS-provided regex functionality by rolling their own, but since this one aims compactness I doubt you plan that. Anyway, I'm glad this issue is resolved, thanks again and have a nice day.

**Comment by raygard** [comment]
@oguz-ismail (and @oliverkwebb ) this is closed but you may be interested to know: while chasing this problem, I took the 23x23 example maze in the AOC (https://adventofcode.com/2023/day/23) and ran it through the various awks. All printed 94/154 except busybox awk, which printed 154/154. So that awk is not only slow, it's defective. I have not tried to find where in 2323.awk it fails.
The hashing routines may need more work, but I also need to keep them short, too.

**Comment by oguz-ismail** [comment]
@raygard busybox awk has many defects, the one you spotted can be reproduced like so:
```
$ busybox awk 'BEGIN{a[5]; f(a)} function f(b){print length(b)}'
0
```
